README.md
[![npm version](https://img.shields.io/npm/v/node-red-contrib-blynk-ws.svg)](https://www.npmjs.com/package/node-red-contrib-blynk-ws)
[![Npm download](https://img.shields.io/npm/dm/node-red-contrib-blynk-ws.svg)](https://www.npmjs.com/package/node-red-contrib-blynk-ws)
[![GitHub stars](https://img.shields.io/github/stars/gablau/node-red-contrib-blynk-ws.svg)](https://github.com/gablau/node-red-contrib-blynk-ws/stargazers)
[![GitHub issues](https://img.shields.io/github/issues/gablau/node-red-contrib-blynk-ws.svg)](https://github.com/gablau/node-red-contrib-blynk-ws/issues)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/gablau/node-red-contrib-blynk-ws/blob/master/LICENSE)

[![NPM](https://nodei.co/npm/node-red-contrib-blynk-ws.png?downloads=true&downloadRank=true&stars=true)](https://www.npmjs.com/package/node-red-contrib-blynk-ws)

__________


# node-red-contrib-blynk-ws
Blynk library implementation for [Node-RED](https://nodered.org/) using WebSockets

## What is Blynk?
Blynk is a platform with iOS and Android apps to control Arduino, Raspberry Pi and the likes over the Internet.
You can easily build graphic interfaces for all your projects by simply dragging and dropping widgets.
If you need more information, please follow these links:

* [Kickstarter campaign](https://www.kickstarter.com/projects/167134865/blynk-build-an-app-for-your-arduino-project-in-5-m/description)
* [Blynk downloads, docs, tutorials](http://www.blynk.cc)
* [Blynk community](http://community.blynk.cc)
* [Facebook](http://www.fb.com/blynkapp)
* [Twitter](http://twitter.com/blynk_app)

## Node-RED blynk Websockets version
This library supports both SSL (wss://) and non secure (ws://) connection to local server and Blynk cloud server.
For local server, wss:// works only if you ve got a certificate installed.
For Blynk cloud server, you can use **ws://blynk-cloud.com/websockets** or with SSL **wss://blynk-cloud.com/websockets** as the server url.

If you installed Node Red globally use this command to install
```npm install --global node-red-contrib-blynk-ws```

### Supported events, commands and widgets

**Events:**
- read
- write
- app (connected/disconected)

**Commands:**
- write
- bridge
- sync (virtual/all)
- set property

**Widgets:**
- emails
- notify
- LCD
- table
- zeRGBa
- styled button

### Changelog

See the [changelog](CHANGELOG.md) for all the details of each releases

### Node-RED palette

![Available Nodes](./docs/img/blynk-palette.png)

### Blynk App Settings

Use Raspberry PI as hardware to access 128 virtual pins or Generic Board for 32.

### How to use

See information panel on every nodes

![Information Panel](./docs/img/readme-info-panel.png)

### Debug

Use the verbose `-v` flag when starting Node-RED to get more information
or use `node-red-log` and enable log on Configuration Node as needed

![Log settings](./docs/img/readme-log-settings.png)


### Compatibility

This library is retrocompatible and can replace **node-red-contrib-blynk-websockets**.
To do this:
- stop node-red 
- install node-red-contrib-blynk-ws `npm install node-red-contrib-blynk-ws`
- remove node-red-contrib-blynk-websockets `npm uninstall node-red-contrib-blynk-websockets`
- edit your flow file `eg: "my-flows.js"` search for _'blynk-websockets-'_ and replace with _'blynk-ws-'_
- start node-red.

__________

### Implementations for other platforms
* [C++, Arduino](https://github.com/blynkkk/blynk-library)
* [Node.js, Espruino, Browsers](https://github.com/vshymanskyy/blynk-library-js)
* [Particle](https://github.com/vshymanskyy/blynk-library-spark)
* [Python, MicroPython](https://github.com/vshymanskyy/blynk-library-python)
* [OpenWrt](https://github.com/vshymanskyy/blynk-library-openwrt)
* [MBED](https://developer.mbed.org/users/vshymanskyy/code/Blynk/)
* [LabVIEW](https://github.com/juncaofish/NI-LabVIEWInterfaceforBlynk)
* [C#](https://github.com/sverrefroy/BlynkLibrary)

### Attributions

The **node-red-contrib-blynk-ws** was born as a fork of **node-red-contrib-blynk-websockets**
     [https://github.com/tzapu/node-red-contrib-blynk-websockets](https://github.com/tzapu/node-red-contrib-blynk-websockets)  
Some javascripts code was derived from **blynk-library-js**:   
     [https://github.com/vshymanskyy/blynk-library-js](https://github.com/vshymanskyy/blynk-library-js) 

### License
This project is released under The MIT License (MIT)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.7.1] - 2018-08-04
### Fixed
- Node - *Sync* - Missing "syncAll" commands

## [0.7.0] - 2018-07-30
### Added
- Node - *Styled Button* -  New node to simplify the use of this new button
- Node - *Set Property* - Implemented "onColor, offColor, onBackColor, offBackColor" property for Styled Button widget (blynk cloud or local server >= v0.36.2)
- Node - *Set Property* - Implemented "fraction" property for Slide widget (blynk cloud or local server >= v0.33.3)
- Node - *Table* - Implemented "select, deselcet and order" commands

### Changed
- Code refactor - Extract blynk method and variable to files in "libs" directory
- Node - *Notify* - Rate limit increased to 5 msg/s (Blynk Library 0.5.3)

### Fixed
- Node - *zeRGBa* - Check if the widget is configured in merge mode, otherwise it will generate an alert
- Missing log on respond illegal command
- Message ID overflow - see Blynk Library JS [PR #44](https://github.com/vshymanskyy/blynk-library-js/pull/44)
- Show protocol warning in debug panel

## [0.6.0] - 2018-05-07
### Added
- Node - *zeRGBa* -  New node to simplify the use of this widget in merge mode

## [0.5.2] - 2018-04-27
### Fixed
- Fix Crash on log Read Event

## [0.5.1] - 2018-04-15
### Fixed
- Fix regular expression for validate server url on configuration node
- Fix display readme

## [0.5.0] - 2018-04-14
### Added
- Blynk protocol - Handle any number of Blynk commands (virtualWrite(), setProperty(), etc.) in a row see [Blynk Library 0.5.0](https://community.blynk.cc/t/beta-blynk-library-v0-5-0/19841)
- Control max lenght of message and max number of commands in a single message 
- Config node - Option to enable "Multiple blynk command in single message"  (blynk cloud or local server >= v0.34.0) 
- Config node - Implemented server proxy configuration
- Config node - Option to log protocol "sync" and "bridge" messages type
- Node - *Bridge* - New node that implements the functionality of sending commands to other blynk devices.
- Node - *Sync* - New node that implements the "syncAll" and "syncVirtual" blynk commands
- Node - *Set Property* - Implemented "step" for Step Widget  (blynk cloud or local server >= v0.32.2)
- Node - *Set Property* - Implemented "url" for Video Straming Widget (blynk cloud or local server >= v0.28.7)
- Node - *Set Property* / *LCD* / *Table* - Send multiple blynk command in a single message
- Node - *Read Event* - Option to receive all pins events
- Node - *Write Event* - Option to receive all pins events
- Node - *Write Event* - Added pin number that generate event in msg.pin variable
- Initial support to [Internationalisation](https://nodered.org/docs/creating-nodes/i18n)

### Changed
- Better handle of configuration node error, now blocks child nodes if some parameter is missing.
- More check on "pin" parameter before send message
- The PING command is only sent if no data has been sent or received in the "heart beat" time
- Hide "Auth token" - from all log message
- Config node - Rename "Private key" to "Auth token" and better help message
- Split every node on single file
- Removed dependency from "DataView" package
- Code clean

### Fixed
- Rightly handles the reception of the BRIDGE type commands.
- Websocket SSL connection error on nodejs > 8.6.0 see ([ws iusse #1227](https://github.com/websockets/ws/issues/1227))

## [0.4.0] - 2018-03-09
### Added
- Handle CONNECT_REDIRECT command see [iusse #5](https://github.com/gablau/node-red-contrib-blynk-ws/issues/5) and [this](https://community.blynk.cc/t/correct-websocket-address-for-blynk-server/22496) 

### Changed
- Server port configuration and help message see [Blynk Library 0.5.1](https://community.blynk.cc/t/new-blynk-library-v0-5-1-is-released-important-for-local-server-owners/22449)

### Fixed
- Invalid HW cmd: "pm" see [iusse #3](https://github.com/gablau/node-red-contrib-blynk-ws/issues/3)
- Protocol Log on message received and retrive right command body

## [0.3.0] - 2017-12-12
### Added
- Node - *Notify* - Implemented queue messages 
- Node - *Write* - Pin mode (Fixed / Dynamic) pass pin number in msg.pin variable
- Node - *Set Property* - Pin mode (Fixed / Dynamic) pass pin number in msg.pin variable

### Changed
- Node - *LCD* - Warning on incorrect use of msg.payload property
- Fixed some log messages

## [0.2.0] - 2017-09-18
### Added
- Debug option to log only specified pins number
- Node - * Table Widgets * to simplify the use of this widget
- Link to official documentation

### Fixed
- Missing logs on read and write event

## 0.1.0 - 2017-09-15
### Added
- Implemented SSL usage on the websocket connection.
- Added Other Messages Type and Status Messages
- Debug option in the configuration node
- Display the number of pins on the status of the nodes,
- Node - * Notify * to send push notifications on the smartphone
- Node - * Set Property * to set the property of any widget
- Node - * App Event * for the connection and disconnect event of the app
- Node - * LCD Widgets * to simplify the use of this widget
- On-line documentation of each node
- Send PROFILE command on login

### Changed
- Blynk Protocol - Rewritten send and receive message functions, with more options, like [vshymanskyy/blynk-library-js](https://github.com/vshymanskyy/blynk-library-js)
- Heartbeat once 10 seconds
- Max virtual pin allowed 128

[Unreleased]: https://github.com/gablau/node-red-contrib-blynk-ws/compare/0.7.1...HEAD
[0.7.1]: https://github.com/gablau/node-red-contrib-blynk-ws/compare/0.7.0...0.7.1
[0.7.0]: https://github.com/gablau/node-red-contrib-blynk-ws/compare/0.6.0...0.7.0
[0.6.0]: https://github.com/gablau/node-red-contrib-blynk-ws/compare/0.5.2...0.6.0
[0.5.1]: https://github.com/gablau/node-red-contrib-blynk-ws/compare/0.5.1...0.5.2

libs/blynk-enum.js
/* blynk enum */

var MsgType = {
	RSP           :  0,
    
	//app commands
	REGISTER      :  1, //"mail pass"
	LOGIN         :  2, //"token" or "mail pass"
	SAVE_PROF     :  3,
	LOAD_PROF     :  4,
	GET_TOKEN     :  5,
	PING          :  6,
	ACTIVATE      :  7, //"DASH_ID"
	DEACTIVATE    :  8, //
	REFRESH       :  9, //"refreshToken DASH_ID"
	GET_GRAPH_DATA : 10,
	GET_GRAPH_DATA_RESPONSE : 11,
    
	//HARDWARE commands
	TWEET         :  12,
	EMAIL         :  13,
	NOTIFY        :  14,
	BRIDGE        :  15,
	HW_SYNC       :  16,
	INTERNAL      :  17, //0x11
	SMS           :  18,
	PROPERTY      :  19,
	HW            :  20, //0x14
    
	//app commands
	CREATE_DASH         : 21,
	UPDATE_DASH         : 22,
	DELETE_DASH         : 23,
	LOAD_PROF_GZ        : 24,
	APP_SYNC            : 25,
	SHARING             : 26,
	ADD_PUSH_TOKEN      : 27,
	EXPORT_GRAPH_DATA   : 28,
	HW_LOGIN            : 29,
	//app sharing commands
	GET_SHARE_TOKEN     : 30,
	REFRESH_SHARE_TOKEN : 31,
	SHARE_LOGIN         : 32,
    
	//app commands
	CREATE_WIDGET       : 33,
	UPDATE_WIDGET       : 34,
	DELETE_WIDGET       : 35,

	//energy commands
	GET_ENERGY          : 36,
	ADD_ENERGY          : 37,

	UPDATE_PROJECT_SETTINGS : 38,

	ASSIGN_TOKEN        : 39,
	GET_SERVER          : 40,
	CONNECT_REDIRECT    : 41,

	CREATE_DEVICE       : 42,
	UPDATE_DEVICE       : 43,
	DELETE_DEVICE       : 44,
	GET_DEVICES         : 45,

	CREATE_TAG          : 46,
	UPDATE_TAG          : 47,
	DELETE_TAG          : 48,
	GET_TAGS            : 49,

	APP_CONNECTED       : 50,

	UPDATE_FACE         : 51,
	//web sockets
	WEB_SOCKETS         : 52,
	EVENTOR             : 53,
	WEB_HOOKS           : 54,
    
	CREATE_APP 					: 55,
	UPDATE_APP 					: 56,
	DELETE_APP 					: 57,
	GET_PROJECT_BY_TOKEN 		: 58,
	EMAIL_QR 					: 59,
	GET_ENHANCED_GRAPH_DATA 	: 60,
	DELETE_ENHANCED_GRAPH_DATA 	: 61,

	GET_CLONE_CODE 				: 62,
	GET_PROJECT_BY_CLONE_CODE 	: 63,

	HARDWARE_LOG_EVENT 			: 64,
	HARDWARE_RESEND_FROM_BLUETOOTH : 65,
	LOGOUT 						: 66,

	CREATE_TILE_TEMPLATE : 67,
	UPDATE_TILE_TEMPLATE : 68,
	DELETE_TILE_TEMPLATE : 69,
	GET_WIDGET			 : 70,
	DEVICE_OFFLINE 		 : 71,
	OUTDATED_APP_NOTIFICATION : 72,
	TRACK_DEVICE 		 : 73,
	GET_PROVISION_TOKEN : 74,

	/*
    //http codes. Used only for stats
    HTTP_IS_HARDWARE_CONNECTED 	: 82,
    HTTP_IS_APP_CONNECTED 		: 83,
    HTTP_GET_PIN_DATA 			: 84,
    HTTP_UPDATE_PIN_DATA 		: 85,
    HTTP_NOTIFY	 				: 86,
    HTTP_EMAIL 					: 87,
    HTTP_GET_PROJECT 			: 88,
    HTTP_QR 					: 89,
    HTTP_GET_HISTORY_DATA 		: 90,
    HTTP_START_OTA 				: 91,
    HTTP_STOP_OTA 				: 92,
    HTTP_CLONE 					: 93,
    HTTP_TOTAL 					: 94,
    */
};

var MsgStatus = {
	OK                          : 200,
	QUOTA_LIMIT_EXCEPTION       : 1,
	ILLEGAL_COMMAND             : 2,
	NOT_REGISTERED              : 3,
	ALREADY_REGISTERED          : 4,
	NOT_AUTHENTICATED           : 5,
	NOT_ALLOWED                 : 6,
	DEVICE_NOT_IN_NETWORK       : 7,
	NO_ACTIVE_DASHBOARD         : 8, 
	INVALID_TOKEN               : 9,
	ILLEGAL_COMMAND_BODY        : 11, 
	GET_GRAPH_DATA_EXCEPTION    : 12,
	NOTIFICATION_INVALID_BODY 	: 13,
	NOTIFICATION_NOT_AUTHORIZED : 14,
	NOTIFICATION_ERROR 			: 15,
	//reserved
	BLYNK_TIMEOUT               : 16,
	NO_DATA_EXCEPTION           : 17,
	//DEVICE_WENT_OFFLINE       : 18, //removed
	SERVER_EXCEPTION            : 19,
	//NOT_SUPPORTED_VERSION       : 20, //removed
	ENERGY_LIMIT                  : 21,
	FACEBOOK_USER_LOGIN_WITH_PASS : 22,
};

exports.MsgType = MsgType;
exports.MsgStatus = MsgStatus;

nodes/blynk-ws-zergba.js
module.exports = function(RED) {
	"use strict";

	var blynkUtil = require('./../libs/blynk-util.js');

	function BlynkZergbaNode(n) {
		RED.nodes.createNode(this, n);
		var node = this;
		this.client = n.client;
		this.pin = n.pin;
		this.nodeType = "zergba";
        
		this.connected_label = RED._("blynk-ws-zergba.status.connected-fixed") + this.pin; 

		this.blynkClient = RED.nodes.getNode(this.client);
		if (this.blynkClient) {
			this.blynkClient.registerInputNode(this);
			this.blynkClient.on("opened", function(n) {
				node.status({
					fill: "yellow",
					shape: "dot",
					text: RED._("blynk-ws-zergba.status.connecting") + n
				});
			});
			this.blynkClient.on("connected", function() {
				node.status({
					fill: "green",
					shape: "dot",
					text: node.connected_label
				});
			});
			this.blynkClient.on("error", function() {
				node.status({
					fill: "red",
					shape: "ring",
					text: "blynk-ws-zergba.status.error"
				});
			});
			this.blynkClient.on("closed", function() {
				node.status({
					fill: "red",
					shape: "ring",
					text: "blynk-ws-zergba.status.disconnected"
				});
			});
		} else {
			this.error(RED._("blynk-ws-zergba.errors.missing-conf"));
		}

		this.on("input", function(msg) {
			if (msg.hasOwnProperty("payload") && node.blynkClient && node.blynkClient.logged) {
				var payload = Buffer.isBuffer(msg.payload) ? msg.payload : RED.util.ensureString(msg.payload);
				var pin = node.pin;
				var newmsg = {};

				if (msg.hasOwnProperty("hex")) {
					var hex = Buffer.isBuffer(msg.hex) ? msg.hex : RED.util.ensureString(msg.hex);
					var color = blynkUtil.hexToRgb(hex);
					newmsg.hex = hex;
					newmsg.r = color.r;
					newmsg.g = color.g;
					newmsg.b = color.b;
					newmsg.rgb = color.r+";"+color.g+";"+color.b;
					newmsg.payload = [color.r, color.g, color.b];
					
				}
				else if (msg.hasOwnProperty("r") && msg.hasOwnProperty("g") && msg.hasOwnProperty("b")) {
					var r = Buffer.isBuffer(msg.r) ? msg.r : RED.util.ensureString(msg.r);
					var g = Buffer.isBuffer(msg.g) ? msg.g : RED.util.ensureString(msg.g);
					var b = Buffer.isBuffer(msg.b) ? msg.b : RED.util.ensureString(msg.b);
					newmsg.hex = blynkUtil.rgbToHex(parseInt(r), parseInt(g), parseInt(b));
					newmsg.r = msg.r;
					newmsg.g = msg.g;
					newmsg.b = msg.b;
					newmsg.rgb = msg.r+";"+msg.g+";"+msg.b;
					newmsg.payload = [msg.r, msg.g, msg.b];
				}
				else {
					var tmpcolor = payload.split("\0");
					if(tmpcolor.length == 3){
						newmsg.hex = blynkUtil.rgbToHex(parseInt(tmpcolor[0]), parseInt(tmpcolor[1]), parseInt(tmpcolor[2]));
						var color = blynkUtil.hexToRgb(newmsg.hex);
						newmsg.r = color.r;
						newmsg.g = color.g;
						newmsg.b = color.b;
						newmsg.rgb = color.r+";"+color.g+";"+color.b;
						newmsg.payload = [color.r, color.g, color.b];

					}
					tmpcolor = payload.split(";");
					if(tmpcolor.length ==3){
						newmsg.hex = blynkUtil.rgbToHex(parseInt(tmpcolor[0]), parseInt(tmpcolor[1]), parseInt(tmpcolor[2]));
						var color = blynkUtil.hexToRgb(newmsg.hex);
						newmsg.r = color.r;
						newmsg.g = color.g;
						newmsg.b = color.b;
						newmsg.rgb = color.r+";"+color.g+";"+color.b;
						newmsg.payload = [color.r, color.g, color.b];
					}
					
				}
				//todo valida valori

				this.status({
					fill: "green",
					shape: "dot",
					text: node.connected_label + " ["+newmsg.r+", "+newmsg.g+", "+newmsg.b+"]",
				});

				node.blynkClient.virtualWrite(pin, newmsg.payload);
				this.send(newmsg);
			}
		});

		this.on("close", function() {
			node.blynkClient.removeInputNode(node);
		});
	}
	
	RED.nodes.registerType("blynk-ws-zergba", BlynkZergbaNode);
};

libs/blynk-lib.js
//blynk enum
var blynkEnum = require('./../libs/blynk-enum.js');
var MsgStatus = blynkEnum.MsgStatus;
var MsgType = blynkEnum.MsgType;
//console.log(MsgType);

//blynk util
var blynkUtil = require('./../libs/blynk-util.js');
var messageToDebugString = blynkUtil.messageToDebugString;
var commandToDebugString = blynkUtil.commandToDebugString;
//var decodeCommand 		 = blynkUtil.decodeCommand;
var blynkHeader 		 = blynkUtil.blynkHeader;
var getStatusByCode = blynkUtil.getStatusByCode;

var srs = require("secure-random-string");
//iniziate SRS
srs({length: 16, alphanumeric: true});

function getTimestamp(){
	return parseInt((new Date).getTime()/1000);	
}

/* ##### BLYNK STUFF ###### */

//library 0.4.7 - 2017-04-09
//library 0.5.1 - 2018-02-20
//library 0.5.2 - 2018-03-04
//library 0.5.3 - 2018-06-11

//Server Version
// 0.23.5 - 2017-04-07
// 0.32.2 - 2018-02-28
// 0.33.3 - 2018-03-20
// 0.34.0 - 2018-04-06
// 0.36.2 - 2018-05-11

// Blynk library constant 
var BLYNK_VERSION = "0.5.3"; //blynk library version
var BLYNK_HEARTBEAT = 10; //seconds
var BLYNK_PROTOCOL_MAX_LENGTH = 32767; //java Short.MAX_VALUE
var BLYNK_MAX_CMD_IN_MESSAGE = 1024; //max command in a single message 

//### PROTOCOL ###

/* build a blynk command */
var blynkCmd = function (msg_type, values, msg_id) {
	var self = this;
	values = values || [""];
	if (msg_type === MsgType.RSP) {
		return blynkHeader(msg_type, (msg_id-1), values);
	}
	else {
		if( self.msg_id > 0xFFFF ) self.msg_id = 1; //max 2 byte
		if( msg_id > 0xFFFF ) msg_id = 1; //max 2 byte
		msg_id = msg_id || (self.msg_id++);
  
		if( typeof values === "string" || typeof values === "number" ) {
			values = [ values ];
		}
    
		var data = values.join("\0");
		return blynkHeader(msg_type, msg_id, data.length) + data;
	}
};

var sendRsp = function (msg_type, msg_id, msg_len, msg_data) {
	var data = msg_data || "";
	var msg = null;
	if (msg_type === MsgType.RSP) {
		msg = this.blynkCmd(msg_type, msg_len, msg_id);
	}
	else {
		msg = this.blynkCmd(msg_type, data, msg_id);
	}

	this.sendMsg(msg);
};

/* send binary message througth websocket */
var sendMsg = function (data) {
	if(data.length<5 || data == undefined) {
		this.log("ERROR sendMsg - No data to send");
		return;
	}
	if (this.dbg_low) {
		this.log("SEND -> " + messageToDebugString(data));
	}
	//convert to Uint8Array
	var rawdata = new Uint8Array(data.length);
	for (var i = 0, j = data.length; i < j; ++i) {
		rawdata[i] = data.charCodeAt(i);
	}
	this.last_activity_out = this.last_heart_beat = getTimestamp();
	//this.log(rawdata);
	this.websocket.send(rawdata);
};

/* Send multiple command message - delete key */
var sendMsgMulti = function (key) {
	this.sendMsg(this.msgList[key]);
	delete this.msgList[key]; //unset key
};

/* Start multiple command message - generate key */
var startMsgMulti = function () {
	var key = srs({length: 16, alphanumeric: true});
	this.msgList[key] = "";
	return key;
};

var processCommand = function (cmd) {

	var RED = this.RED;

	if (!this.logged) {
		if (cmd.type === MsgType.RSP && cmd.msgId === 1) {
			if (cmd.len === MsgStatus.OK || cmd.len === MsgStatus.ALREADY_REGISTERED) {
				this.log("Client logged");
				this.logged = true;
				this.sendInfo();
				this.emit("connected", "");
			}
			else if(cmd.len === MsgStatus.BLYNK_INVALID_TOKEN) {
				this.log("Invalid auth token");
			}
			else if(cmd.len === MsgStatus.NOT_AUTHENTICATED) {
				this.log("Not autenticated");
			}
			else {
				this.log("Connect failed. code: "+cmd.len);
			}
		}
		if(cmd.type === MsgType.RSP && cmd.len === MsgStatus.NOT_AUTHENTICATED) {
			this.log("Not autenticated");
		}
		if(cmd.type === MsgType.CONNECT_REDIRECT) {
			//handle server redirect 
			var schema = "ws://";
			var port = 80;
			if (this.path.startsWith("wss://")) {
				schema = "wss://";
				port = 443;
			}
			var values = cmd.body.split("\0");
			var serverip = values[0];
			if(values[1]>0) port = values[1];
			var newpath = schema + serverip + ":" + port + "/websockets";
			this.path = newpath; 
			this.warn(RED._("Connection redirecting to:  " + newpath));
		}
	} else { //
		this.last_activity_in = getTimestamp();
		switch (cmd.type) {
		case MsgType.RSP:
			if (cmd.status !== MsgStatus.OK) { //handle not ok response message
				var error=false;
				for (var k in MsgStatus) {
					if (MsgStatus.hasOwnProperty(k)) {
						if(cmd.status == MsgStatus[k]) error=true;
					}
				}
				if(error) { //handle know error 
					switch(cmd.status){
						case MsgStatus.NOT_AUTHENTICATED: 
							this.log("Not autenticated");
							break;
						case MsgStatus.BLYNK_INVALID_TOKEN: 
							this.log("Invalid auth token");
							break;
						case MsgStatus.ILLEGAL_COMMAND_BODY: 
							//this.log("Illegal command body");
							break;	
						default:
							this.warn(RED._("Response Error: "+getStatusByCode(cmd.status)));
							break;
					}
				}
				else { 
					this.warn(RED._("Unhandled RSP type: " + commandToDebugString(cmd)));
				}
			}
			break;
		case MsgType.LOGIN:
		case MsgType.PING:
			this.sendRsp(MsgType.RSP, this.msg_id, MsgStatus.OK);
			break;
		case MsgType.HW:
		case MsgType.BRIDGE:
			//this.warn(cmd.typeString+" cmd: " + JSON.stringify(cmd));
			switch (cmd.operation) {
			//input nodes
			case "dw":
			case "vw":
				this.handleWriteEvent(cmd);
				break;
			case "dr":
			case "vr":
				this.handleReadEvent(cmd);
				break;
			case "pm":
				// skip message "pin mode"
				break;
			default:
				this.warn(RED._("Invalid "+cmd.typeString+" cmd: " + commandToDebugString(cmd)));
				//this.sendRsp(MsgType.RSP, this.msg_id, MsgStatus.ILLEGAL_COMMAND);
				this.sendRspIllegalCmd(this.msg_id);
			}
			break;
        
		case MsgType.INTERNAL:
			switch (cmd.body) {
			//app event node
			case "acon":
			case "adis":
				this.handleAppEvent(cmd.body);
				break;
			default:
				this.warn(RED._("Invalid INTERNAL cmd: " + commandToDebugString(cmd)));
                //this.sendRsp(MsgType.RSP, this.msg_id, MsgStatus.ILLEGAL_COMMAND);
			}
			break;
		case MsgType.GET_TOKEN:
			this.sendRsp(MsgType.GET_TOKEN, this.msg_id, this.key.length, this.key);
			break;
		case MsgType.LOAD_PROF:
			//this.sendRsp(MsgType.LOAD_PROF, this.msg_id, profile.length, self.profile);
			break;
		case MsgType.DEBUG_PRINT:
			this.log("Server: " + cmd.body);
			break;
		case MsgType.REGISTER:
		case MsgType.SAVE_PROF:
		case MsgType.ACTIVATE:
		case MsgType.DEACTIVATE:
		case MsgType.REFRESH:
			// skip this message types
			break;
		default:
			this.warn(RED._("Invalid header type: " + commandToDebugString(cmd)));
			//this.sendRsp(MsgType.RSP, this.msg_id, MsgStatus.ILLEGAL_COMMAND);
			this.sendRspIllegalCmd(this.msg_id);

		}
	}
};


//### PROTOCOL COMMAND ###

/* send login message */
var login = function(token) {
	if(this.dbg_all){
		this.log("login -> " + String("********************************" + token.slice(-5)).slice(-32));
	}
	this.sendMsg(this.blynkCmd(MsgType.LOGIN, token, 1));
};

/* send Illegal command response message */
var sendRspIllegalCmd = function (msg_id) {
	this.sendRsp(MsgType.RSP, msg_id, MsgStatus.ILLEGAL_COMMAND);
};

/* send info message */
var sendInfo = function () {
	var info = [
		"ver", BLYNK_VERSION, 
		"h-beat", BLYNK_HEARTBEAT, 
		"buff-in", BLYNK_PROTOCOL_MAX_LENGTH, 
		"dev", "node-red", 
		"con", "Blynk-ws",
		"build", this.LIBRARY_INFO,
	];
	this.msg_id++;
	this.sendMsg(this.blynkCmd(MsgType.INTERNAL, info));
};

/* send ping message */
var ping = function() {
	if(this.dbg_all || this.dbg_read){
		this.log("ping");
	}
	this.last_heart_beat = getTimestamp();
	this.sendMsg(this.blynkCmd(MsgType.PING));
};

/* send syncAll message */
var syncAll = function(msgkey) {
	if(this.dbg_all || this.dbg_sync){
		this.log("syncAll: -> all");
	}
	var msg = this.blynkCmd(MsgType.HW_SYNC);
	if(this.multi_cmd && msgkey != undefined) 
		this.msgList[msgkey] = this.msgList[msgkey] + msg;
	else this.sendMsg(msg);
};


/* send syncVirtual message */
var syncVirtual = function(vpin, msgkey) {
	if(this.dbg_all || this.dbg_sync || this.isLogPin(vpin)){
		this.log("syncVirtual: -> " + JSON.stringify(["vr", vpin]));
	}
	var msg = this.blynkCmd(MsgType.HW_SYNC , ["vr", vpin]);
	if(this.multi_cmd && msgkey != undefined) 
		this.msgList[msgkey] = this.msgList[msgkey] + msg;
	else this.sendMsg(msg);
};

/* send virtualWrite message */
var virtualWrite = function(vpin, val, msgkey) {
	if(this.dbg_all || this.dbg_write || this.isLogPin(vpin)){
		this.log("virtualWrite: -> " + JSON.stringify(["vw", vpin].concat(val)));
	}
	var msg = this.blynkCmd(MsgType.HW, ["vw", vpin].concat(val));
	if(this.multi_cmd && msgkey != undefined) 
		this.msgList[msgkey] = this.msgList[msgkey] + msg;
	else this.sendMsg(msg);
};

/* send setProperty message - set msgkey form multimple command message*/
var setProperty = function(vpin, prop, val, msgkey) {
	if(this.dbg_all  || this.dbg_prop || this.isLogPin(vpin)){
		this.log("setProperty -> " + JSON.stringify([vpin, prop].concat(val)));
	}
	var msg = this.blynkCmd(MsgType.PROPERTY, [vpin, prop].concat(val));
	if(this.multi_cmd && msgkey != undefined) 
		this.msgList[msgkey] = this.msgList[msgkey] + msg;
	else this.sendMsg(msg);
};

var sendEmail = function(to, subject, message) {
	if(this.dbg_all || this.dbg_mail){
		this.log("sendEmail -> " + JSON.stringify([to, subject, message]));
	}
	this.sendMsg(this.blynkCmd(MsgType.EMAIL, [to, subject, message]));
};

var sendNotify = function(message) {
	if(this.dbg_all || this.dbg_notify){
		this.log("sendNotify -> " + JSON.stringify([message]));
	}
	this.sendMsg(this.blynkCmd(MsgType.NOTIFY, [message]));
};

/* send bridgeSetAuthToken message */
var bridgeSetAuthToken = function(bpin, token) {
	if(this.dbg_all || this.dbg_bridge || this.isLogPin(bpin)){
		var logtoken = String("********************************" + token.slice(-5)).slice(-32);
		this.log("bridgeSetAuthToken: -> " + JSON.stringify([bpin, "i", logtoken]));
	}
	var msg = this.blynkCmd(MsgType.BRIDGE, [bpin, "i", token]);
	this.sendMsg(msg);
};

/* send bridgeVirtualWrite message */
var bridgeVirtualWrite = function(bpin, pin, val) {
	if(this.dbg_all || this.dbg_bridge || this.isLogPin(bpin)){
		this.log("bridgeVirtualWrite: -> " + JSON.stringify([bpin, "vw", pin].concat(val)));
	}
	var msg = this.blynkCmd(MsgType.BRIDGE, [bpin, "vw", pin].concat(val));
	this.sendMsg(msg);
};

/* send bridgeAnalogWrite message */
var bridgeAnalogWrite = function(bpin, pin, val) {
	if(this.dbg_all || this.dbg_bridge || this.isLogPin(bpin)){
		this.log("bridgeAnalogWrite: -> " + JSON.stringify([bpin, "aw", pin].concat(val)));
	}
	var msg = this.blynkCmd(MsgType.BRIDGE, [bpin, "aw", pin].concat(val));
	this.sendMsg(msg);
};

/* send bridgeDigitalWrite message */
var bridgeDigitalWrite = function(bpin, pin, val) {
	if(this.dbg_all || this.dbg_bridge || this.isLogPin(bpin)){
		this.log("bridgeDigitalWrite: -> " + JSON.stringify([bpin, "dw", pin].concat(val)));
	}
	var msg = this.blynkCmd(MsgType.BRIDGE, [bpin, "dw", pin].concat(val));
	this.sendMsg(msg);
};

exports.BLYNK_VERSION = BLYNK_VERSION;
exports.BLYNK_HEARTBEAT = BLYNK_HEARTBEAT;
exports.BLYNK_PROTOCOL_MAX_LENGTH = BLYNK_PROTOCOL_MAX_LENGTH;
exports.BLYNK_MAX_CMD_IN_MESSAGE = BLYNK_MAX_CMD_IN_MESSAGE; 

//protocol 
exports.blynkCmd = blynkCmd;
exports.sendRsp = sendRsp;
exports.sendMsg = sendMsg;
exports.sendMsgMulti = sendMsgMulti;
exports.startMsgMulti = startMsgMulti;
exports.processCommand = processCommand;


//protocol command
exports.login = login;
exports.sendInfo = sendInfo;
exports.sendRspIllegalCmd = sendRspIllegalCmd;
exports.ping = ping;
exports.syncAll = syncAll;
exports.syncVirtual = syncVirtual;
exports.virtualWrite = virtualWrite;
exports.setProperty = setProperty;
exports.sendEmail = sendEmail;
exports.sendNotify = sendNotify;
exports.bridgeSetAuthToken = bridgeSetAuthToken;
exports.bridgeVirtualWrite = bridgeVirtualWrite;
exports.bridgeAnalogWrite = bridgeAnalogWrite;
exports.bridgeDigitalWrite = bridgeDigitalWrite;
